windows/devices/uwb/UwbConnector.cxx

```diff
@@ -913,99 +922,68 @@ UwbConnector::RegisterSessionEventCallbacks(uint32_t sessionId, std::weak_ptr<::
 bool
 UwbConnector::CallbacksPresent()
 {
-    return (not m_deviceEventCallbacksIdMap.empty()) or (not m_sessionEventCallbacksIdMap.empty());
+    return (not m_sessionEventCallbacks.empty()) or (not m_deviceEventCallbacks.empty());
 }
 
 void
 UwbConnector::DeregisterEventCallback(std::weak_ptr<::uwb::RegisteredCallbackToken> token)
 {
-    auto tok = token.lock();
-    if (not tok) {
+    auto tokenShared = token.lock();
+    if (not tokenShared) {
         return;
     }
     std::lock_guard eventCallbacksLockExclusive{ m_eventCallbacksGate };
 
-    DeregisterEventCallback(*tok);
-}
-
-void
-UwbConnector::DeregisterEventCallback(::uwb::RegisteredCallbackToken token)
-{
-    auto callbackId = token.callbackId;
-    auto isDeviceEventCallback = token.isDeviceEventCallback;
-
-    // remove the token from the list of tokens
-    for (auto it = std::cbegin(m_tokens); it != std::cend(m_tokens); it++) {
-        if ((*it)->callbackId == callbackId) {
-            m_tokens.erase(it);
-            break;
-        }
-    }
-
-    if (isDeviceEventCallback) {
-        // first remove the id from the id map
-        auto result = m_deviceEventCallbacksIdMap.find(callbackId);
-        if (result == std::cend(m_deviceEventCallbacksIdMap)) {
-            return;
+    auto deviceCallback = std::dynamic_pointer_cast<::uwb::RegisteredDeviceCallbackToken>(tokenShared);
+    if (deviceCallback) {
+        // treat it as a device callback
+        auto tokenIt = std::find_if(std::cbegin(m_deviceEventCallbacks), std::cend(m_deviceEventCallbacks), [deviceCallback](const auto& token) {
+            return token.get() == deviceCallback.get();
+        });
+        if (tokenIt == std::cend(m_deviceEventCallbacks)) {
+            return; // no associated token found, bail
         }
-        m_deviceEventCallbacksIdMap.erase(result);
+        m_deviceEventCallbacks.erase(tokenIt);
     } else {
-        // first remove the id from the id map
-        auto result = m_sessionEventCallbacksIdMap.find(callbackId);
-        if (result == std::cend(m_sessionEventCallbacksIdMap)) {
-            return;
-        }
-        m_sessionEventCallbacksIdMap.erase(result);
-
-        // now remove the id from the list of m_sessionIdToCallbackIdsMap
-        auto& [_, pair] = *result;
-        auto& [sessionId, callbackWeak] = pair;
+        auto sessionCallback = dynamic_pointer_cast<::uwb::RegisteredSessionCallbackToken>(tokenShared);

```

Comment:
Right now `sessionCallback` will always be non-`nullptr` since we only have 2 derived types, but we should guard against the case that ever changes. So, let's add an additional check here:
```suggestion
        auto sessionCallback = dynamic_pointer_cast<::uwb::RegisteredSessionCallbackToken>(tokenShared);
        if (sessionCallback == nullptr) {
            throw std::runtime_error("invalid callback type used, this is a bug!");
        }
```